### PR TITLE
Implement tokenex regular expression matching and add tests.

### PR DESCRIPTION
**Pull request by yuslepukhin**

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -238,8 +239,10 @@ set(ONNXRUNTIME_INCLUDE_DIR ${REPO_ROOT}/include/onnxruntime)
 
 add_subdirectory(external/date EXCLUDE_FROM_ALL)
 add_subdirectory(external/gsl EXCLUDE_FROM_ALL)
+add_subdirectory(external/re2 EXCLUDE_FROM_ALL)
 add_library(date ALIAS tz)
 add_library(gsl ALIAS GSL)
+add_library(re2 ALIAS re2)
```
Comment (snnn) [inline cmake/CMakeLists.txt]: This line is not needed.

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 2]
```diff
@@ -1,19 +1,49 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "tokenizer.h"
 #include "onnx/defs/schema.h"
 #include "core/common/common.h"
```
Comment (snnn) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc]: see : [onnxruntime/core/graph/onnx_protobuf.h](/onnxruntime/core/graph/onnx_protobuf.h)

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 3]
```diff
@@ -209,38 +240,53 @@ Tokenizer::Tokenizer(const OpKernelInfo& info) : OpKernel(info) {
...
-  ORT_ENFORCE(status.IsOK(), "attribute separators is not set");
-  ORT_ENFORCE(!separators.empty(), "Requires at least one separator");
+  std::string tokenexp;
+  status = info.GetAttrs("separators", separators);
+  if (!status.IsOK()) {
+    status = info.GetAttr("tokenexp", &tokenexp);
+    ORT_ENFORCE(status.IsOK(), "Either one of the separators OR tokenexp attributes required but none is set");
+    ORT_ENFORCE(!tokenexp.empty(), "Expecting a non-empty tokenexp");
+  } else {
+    ORT_ENFORCE(!separators.empty(), "Expect at least on item within separators");
```
Comment (askhade) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc]: nit pick : typo on -> one

File: onnxruntime/core/platform/windows/debug_alloc.cc [hunk 4]
```diff
@@ -210,7 +210,7 @@ Memory_LeakCheck::~Memory_LeakCheck() {
     //     empty_string = new string;
     //     empty_named_groups = new std::map<string, int>;
     //     empty_group_names = new std::map<int, string>; });
-    if (string.find("re2::RE2::Init") == std::string::npos &&
+    if (string.find("RtlRunOnceExecuteOnce") == std::string::npos &&
```
Comment (snnn) [inline onnxruntime/core/platform/windows/debug_alloc.cc]: It's a Windows Symbol. Are you sure you want to do this?

File: .gitmodules [hunk 5]
```diff
@@ -25,3 +25,6 @@
 [submodule "cmake/external/onnx-tensorrt"]
 	path = cmake/external/onnx-tensorrt
 	url = https://github.com/stevenlix/onnx-tensorrt.git
+[submodule "cmake/external/re2"]
```
[the remark this answers is not in the document]
Comment (snnn) [reply .gitmodules]: Hi @pranavsharma, let's do it in another PR.

File: onnxruntime/core/platform/windows/debug_alloc.cc (hunk 4 above)
Comment (yuslepukhin) [reply onnxruntime/core/platform/windows/debug_alloc.cc]: Yes, it is a windows source code :)

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 6]
```diff
@@ -447,6 +493,133 @@ Status Tokenizer::SeparatorTokenize(OpKernelContext* ctx,
...
+    const char* data = curr_input->data();
+    for (const auto& token : row) {
+#ifdef _DEBUG
+      auto s_len = curr_input->length();
+      assert(token.second > 0);
+      assert(token.first < s_len);
+      assert(token.first + token.second <= s_len);
+#endif
+      (output_data + output_index)->assign(data + token.first, token.second);
+      ++output_index;
```
Comment (wschin) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc]: I am not sure if it works with multi-byte words in UTF-8.

**Comment by wschin** [comment]
>       (output_data + output_index)->assign(&start_text, 1);

Better to use size of `start_text`?

---
Refers to: onnxruntime/contrib_ops/cpu/tokenizer.cc:333 in 59f4138. [](commit_id = 59f4138b8fa52c2271c2904704130f59dd8beb30, deletion_comment = False)

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 7]
```diff
@@ -447,6 +493,133 @@ Status Tokenizer::SeparatorTokenize(OpKernelContext* ctx,
...
+    }
+    if (mark_) {
+      (output_data + output_index)->assign(&end_text, 1);
+      ++output_index;
+    }
+    const size_t pads = max_tokens - (mark_ * 2) - row.size();
+    for (size_t p = 0; p < pads; ++p) {
+      *(output_data + output_index) = pad_value_;
+      ++output_index;
```
Comment (wschin) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc]: >output_index [](start = 8, length = 12)

I am not quite sure `pad_value_` is always an 1-byte UTF-8 string.

File: onnxruntime/contrib_ops/cpu/tokenizer.cc [hunk 8]
```diff
@@ -482,7 +655,11 @@ Status Tokenizer::Compute(OpKernelContext* ctx) const {
   if (char_tokenezation_) {
     s = CharTokenize(ctx, N, C, input_dims);
   } else {
-    s = SeparatorTokenize(ctx, N, C, input_dims);
+    if (regex_ != nullptr) {
+      s = ExpressionTokenize(ctx, N, C, input_dims);
```
Comment (wschin) [inline onnxruntime/contrib_ops/cpu/tokenizer.cc]: >N, C [](start = 34, length = 4)

Is this (N, C) computed in UTF-8 format or `char`?

File: onnxruntime/contrib_ops/cpu/tokenizer.cc (hunk 6 above)
Comment (yuslepukhin) [reply onnxruntime/contrib_ops/cpu/tokenizer.cc]: It works

File: onnxruntime/contrib_ops/cpu/tokenizer.cc (hunk 8 above)
Comment (yuslepukhin) [reply onnxruntime/contrib_ops/cpu/tokenizer.cc]: Dimensions describe the number of std::string objects and do not count chars.

File: onnxruntime/contrib_ops/cpu/tokenizer.cc (hunk 7 above)
Comment (yuslepukhin) [reply onnxruntime/contrib_ops/cpu/tokenizer.cc]: pad_value_ is instance of `std::string` object and the assignment invokes `operator=`

**Pull request merged**